README
hlbr/README.en

hlbr/CHANGELOG
2011-09-07

 * Merged a patch sent by Aurelio Costa (arcosta@gmail.com)
   This patch is a complement for Bertinatto's patch (Merged
   by PAJ);

2011-02-10

 * Merged a patch sent by Fábio Bertinatto (fabiojrb@gmail.com)
   This patch aims to change the HLBR uid from root to a less
   privileged one as soon as files are open and interfaces are
   set as promiscuous (Merge by PAJ);

2010-

 * Many changes that were not documented (PAJ);
 * Most of these changes were related to parallelism;
 * Improvement of ~40% in HLBR performance;

2009-11-03

 * Added the possibility to control the thread's subsystem
   through HLBR's configuration file; (PAJ)
 * hlbr.config.pt_BR update with information about the
   feature cited above; (PAJ)
 * Refactoring of some functions in engine/main_loop.c; (PAJ)
 * BUGFIX: Refactoring of the IPDefrag decoder. No more
   race conditions exists. Now, the code is a little bit
   more efficient; (PAJ)
 * BUGFIX: HLBR was returning 1 instead of zero on
   sucessful shutdowns. (PAJ)

2009-09-03

 * Actualy implementation of MacFilter routing method (PAJ)
 * BUGFIX: SchedulePacket(int PacketSlot) rewrited (PAJ)

2009-07-02

 * added debug situation to screen warnings about ID, Proto and adding slot in
   fragment identification in decoders/decode_ip_defrag.c (ERI)

2009-03-08

 * Improved paralelism (PAJ)
 * Implementation of thread-safe data structures (PAJ)
 * Actualy implementation of SBridge routing method (PAJ)
 * Got rid of hlbr_mutex_* functions (PAJ)
 * Got rid of non-threaded running mode (PAJ)

2008-11-08

 * Thread for keeping of logfiles (ABA)

2008-09-28

 * BUGFIX: Packet allocation status in GetEmptyPacket at packet.c (PAJ)
 * BUGFIX: Memory leak in the new decoder data indexing schema (PAJ)
 * BUGFIX: Log file being open two times and not been kept open (PAJ)
 * PacketMutex optmization (PAJ)
 * New decoder data indexing schema (ABA)
 * Multi-threaded decoding (PAJ)

2008-07-24  Version: 1.6

 * added README.tcpdump (ERI)
 * Tests arguments resized (ABA,PAJ)
 * improved pid file process (ABA)
 * added pid file (PAJ)
 * rebuilt init.d script to make it compliant with Debian
   (using start-stop-daemon, if it exists, to stop HLBR) (ERI)

2008-07-18  Version: 1.6 RC1

 * BUGFIX: IPDefrag now working! (PAJ)
 * BUGFIX: JTree issues (PAJ)
 * added decoder especific memory deallocation functions (PAJ)
 * added install-daemon in Makefile.in (for tests only) (ERI)
 * changed some rules (ERI)
 * revision and tests (ABA,ERI)

2008-06-18  Version: 1.5 RC2

 * uri test moved to http test (still testing) (PAJ)
 * added test: http method (PAJ)
 * added uri decoder to engine (PAJ)
 * now decoders may have config options (PAJ)
 * new (and more efficient!) http header detection (PAJ)
 * updated test_ethernet_{src,dst}.c (PAJ)
 * updated README.{en,pt_BR} (ERI)
 * changed tcpdump header from /var/lib/hlbr to /usr/lib/hlbr (ERI)
 * added description and force-reload option in init.d to make it compliant
   with Debian and LSB (ERI)
 * new rules (ERI)
 * revision and tests (ABA,ERI)

2008-05-29  Version: 1.5 RC1

 * new rules (ERI)
 * the POSIX REGEX was replaced by Perl REGEX (PAJ)
 * officialy we support GNU/Linux only now (ERI)
 * added the COPYING file (ERI)
 * removed the LICENSE.{en,pt_BR} files (ERI)
 * changed man/hlbr.8.gz file to man/hlbr.8 (ERI)
 * fixed a bug in Makefile: the manpage is being installed now
 * optimized the Makefile.in (ERI)
 * little updates and fixies in manpage (ERI)
 * removed CHANGELOG.pt_BR and TODO.pt_BR files (ERI)
 * moved {CHANGELOG,TODO}.en files to CHANGELOG and TODO (ERI)
 * the changelog format was remade (ERI)
 * new detection test: tcp uri (in testing phase...) (PAJ)
 * updated README.{en,pt_BR} and TEAM files (ERI)

2006-05-18  Version: 1.1

 * new rules (ERI)

2006-03-05  Version: 1.0

 * new tests: implementation of regular expressions with tcp regex and
   udp regex (MOR,ROV)
 * new tests: tcp listcontent and tcp listnocase (to read files lists) (ABA)
 * new rules using regular expressions (ERI)
 * corrected bug: nocase and content had detection fails (ABA)

2006-02-09  Version: 0.2

 * corrected bug: HLBR didn't start correctly out of the directory of
   configurations and rules (GAB)
 * corrected bug: .dump files getting corrupted (ABA,ERI)
 * on-line manual created (ERI)
 * new rules created (ERI)
 * new log rotating system (ERI)
 * installation system altered (ERI)

2005-12-16  Version: 0.1

 * retreat some of the comments let into the code (ABA)
 * created one action and one log only for e-mail virus (ABA)
 * new rules and "tunning" of some rules already existing (ERI)
 * implementation of initialization by files in /etc/init.d (ERI)
 * log rotate implementation (ERI)

2005-12-05  Version: 0.1 RC2

 * hlbr.config revised (ERI)
 * README.pt_BR revised and more stuff added (ERI)
 * README.en created (ABA,ERI)
 * hlbr.config.en created (ABA,ERI)
 * created: make install, installen and uninstall (ERI)
 * first source code standard revision (ABA)
 * new rules (ERI)

2005-11-29  Version: 0.1 RC1

 * initial version (from Hogwash - hogwash.sf.net) (ABA)



------------------
(ABA) - Andre Bertelli - Brazil
(ERI) - Eriberto - Brazil
(GAB) - Gabriel Arellano - Argentina
(MOR) - Morpheuz - Brazil
(PAJ) - Pedro Arthur - Brazil
(ROV) - Rodrigo Vivi - Brazil


hlbr/engine/bits.c
//#define DEBUG

#include "bits.h"
#include "hlbr.h"
#include <stdio.h>

/***************************************************
* Retrieve the value of a bit
***************************************************/
inline int GetBit(unsigned char* BitField, int BitFieldLen, int BitNum){
	char	byte;
	
	DEBUGPATH;

	if (BitNum > BitFieldLen-1) return FALSE;

	byte=BitField[BitNum/8];
	switch (BitNum%8){
	case 0:
		return byte & 0x80;
	case 1:
		return byte & 0x40;
	case 2:
		return byte & 0x20;
	case 3:
		return byte & 0x10;
	case 4:
		return byte & 0x08;
	case 5:
		return byte & 0x04;
	case 6:
		return byte & 0x02;
	case 7:
		return byte & 0x01;
	}	
	
	return FALSE;
}

/***************************************************
* Set the value of a bit
***************************************************/
inline void SetBit(unsigned char* BitField, int BitFieldLen, int BitNum, char Value){
	unsigned char*	byte;

	DEBUGPATH;
#ifdef DEBUG
	printf("BitFieldLen %i BitNum %i Value %c\n", BitFieldLen, BitNum, Value);
#endif
	
	byte=&BitField[BitNum/8];
	
	if (Value){
		switch (BitNum%8){
		case 0:
			*byte |= 0x80;
			break;
		case 1:
			*byte |= 0x40;
			break;
		case 2:
			*byte |= 0x20;
			break;
		case 3:
			*byte |= 0x10;
			break;
		case 4:
			*byte |= 0x08;
			break;
		case 5:
			*byte |= 0x04;
			break;
		case 6:
			*byte |= 0x02;
			break;
		case 7:
			*byte |= 0x01;
			break;
		}	
	}else{
		switch (BitNum%8){
		case 0:
			*byte &= 0x7F;
			break;
		case 1:
			*byte &= 0xBF;
			break;
		case 2:
			*byte &= 0xDF;
			break;
		case 3:
			*byte &= 0xEF;
			break;
		case 4:
			*byte &= 0xF7;
			break;
		case 5:
			*byte &= 0xFB;
			break;
		case 6:
			*byte &= 0xFD;
			break;
		case 7:
			*byte &= 0xFE;
			break;
		}	
	}	
}

/***************************************************
* Set the value of a range of bits
***************************************************/
inline void SetBits(unsigned char* BitField, int BitFieldLen, int StartBit, int EndBit, char Value){

  DEBUGPATH;

}

/***************************************************
* Calculate the bitwise NOT AND of the two bitfields
***************************************************/
inline void NotAndBitFields(unsigned char* BitField1, unsigned char* BitField2, unsigned char* TargetBitField, int BitFieldLen){
	unsigned int*	IntField1;
	unsigned int*	IntField2;
	unsigned int*	IntTarget;
	register int	i;
	register int	len;
	
	DEBUGPATH;

	IntField1=(unsigned int*)BitField1;
	IntField2=(unsigned int*)BitField2;
	IntTarget=(unsigned int*)TargetBitField;
	
	len = (BitFieldLen/32)+1;
	for (i=0;i<len;i++)
		IntTarget[i]=(IntField2[i]^0xFFFFFFFF) & IntField1[i];
}

/***************************************************
* Calculate the bitwise AND of the two bitfields
***************************************************/
inline void AndBitFields(unsigned char* BitField1, unsigned char* BitField2, unsigned char* TargetBitField, int BitFieldLen){	

  DEBUGPATH;

}

/***************************************************
* Calculate the bitwise OR of the two bitfields
***************************************************/
inline void OrBitFields(unsigned char* BitField1, unsigned char* BitField2, unsigned char* TargetBitField, int BitFieldLen){

  DEBUGPATH;

}

/***************************************************
* Calculate the bitwise OR of the two bitfields
* This is the slow way.  Finish the faster way later.
***************************************************/
int	CountBits(unsigned char* BitField, int BitFieldLen){
	int	i;
	int	count;
	
	DEBUGPATH;

	count=0;
	for (i=0;i<BitFieldLen;i++)
		if (GetBit(BitField, BitFieldLen, i)) count++;
		
	return count;
}

/***************************************************
* return the number of bits that are true
* TODO: finish this
***************************************************/
int	CountBitsNot(unsigned char* BitField, int BitFieldLen){
	int i;
	int	count;
	
	DEBUGPATH;
	
	count = 0;
	for (i=0;i<(BitFieldLen/8);i++){
		switch (BitField[i]){
		case 0:
			count+=0;
			break;
		case 1:
		case 2:
		case 4:
		case 8:
		case 16:
		case 32:
		case 64:
		case 128:
			count+=1;
			break;
		case 3:
		case 5:
		case 6:
		case 9:
		case 10:
		case 12:
		case 17:
		case 18:
		case 20:
		case 24:
		case 33:
		case 34:
		case 36:
		case 40:
		case 65:
		case 129:
			count+=2;
			break;
		case 7:	
		case 11:
		case 13:
		case 14:
		case 19:
		case 21:
		case 22:	
		case 25:
		case 26:
		case 28:
		case 35:
		case 37:
		case 38:	
		case 41:
		case 42:
		case 44:
			count+=3;
			break;	
		case 15:
		case 23:
		case 27:
		case 29:
		case 30:
		case 39:
		case 43:
			count+=4;
			break;
		case 31:
			count+=5;	
		}
	}

	return count;
}


/************************************************************
* Returns true if a bitfield is empty
*************************************************************/
int BitFieldIsEmpty(unsigned char* BitField, int BitFieldLen){
	int*	Field;
	int		i,j;
	
	DEBUGPATH;

	Field=(int*)BitField;

	for (i=0;i<BitFieldLen/32;i++){
		if (Field[i]!=0) return FALSE;
	}
	
	for (j=i*32;j<BitFieldLen;j++){
		if (GetBit(BitField, BitFieldLen, j)) return FALSE;
	}
	
	return TRUE;
}

#ifdef DEBUG
#undef DEBUG
#endif
